File: recorder/usage.go
package recorder

import (
	"sort"
	"strings"

	"google.golang.org/genai"

	pb "github.com/techbridgeinnovation/agentpulse/recorder/pb/metering"
)

// The reporting conventions a set of token counts can arrive in.
//
// Which one a record is in decides how its numbers are read, and it is not the same question as who billed: Claude served through Vertex is billed by Google and reports Vertex-shaped usage. Stated on the record because providers disagree about what their own headline numbers contain, and reading one convention as another charges for the same token twice.
const (
	FormatVertex          = "VERTEX"
	FormatAnthropic       = "ANTHROPIC"
	FormatOpenAIChat      = "OPENAI_CHAT"
	FormatOpenAIResponses = "OPENAI_RESPONSES"

	// FormatPerplexity is OpenAI's chat field names with Perplexity's own counts beside them, such as the citation tokens its search charges for.
	FormatPerplexity = "PERPLEXITY"
)

// The providers that bill for a call, in the vocabulary the rate card prices against.
//
// Constants for the ones we call today, and a plain string on the wire so that calling a provider nobody has called before needs no contract change and no release of this library.
const (
	ProviderVertexAI   = "VERTEX_AI"
	ProviderAnthropic  = "ANTHROPIC"
	ProviderOpenAI     = "OPENAI"
	ProviderPerplexity = "PERPLEXITY"
)

// reportedQuantities turns what a caller read off its provider into what the record carries.
//
// Sorted by unit, because a record whose fields land in a different order on every call is one nothing can be diffed against, and comparing two recordings of the same call is how a convention gets checked.
func reportedQuantities(reported map[string]int64) []*pb.ReportedQuantity {
	if len(reported) == 0 {
		return nil
	}
	units := make([]string, 0, len(reported))
	for unit := range reported {
		if strings.TrimSpace(unit) != "" && reported[unit] != 0 {
			units = append(units, unit)
		}
	}
	sort.Strings(units)

	quantities := make([]*pb.ReportedQuantity, 0, len(units))
	for _, unit := range units {
		quantities = append(quantities, &pb.ReportedQuantity{Unit: unit, Quantity: reported[unit]})
	}
	return quantities
}

// BilledByOf names the provider that billed for a call.
//
// Takes what the adopter stated, and falls back to reading the enum the same thing used to be stated in, so that an agent configured before `BilledBy` existed keeps attributing its spend to the right provider without being touched.
//
//nolint:staticcheck // Reading the deprecated enum is the whole job of this function.
func BilledByOf(billedBy string, provider pb.Activity_Provider) string {
	if billedBy != "" {
		return billedBy
	}
	switch provider {
	case pb.Activity_ANTHROPIC:
		return ProviderAnthropic
	case pb.Activity_OPENAI:
		return ProviderOpenAI
	case pb.Activity_PERPLEXITY:
		return ProviderPerplexity
	default:
		return ProviderVertexAI
	}
}

// ReportedFromGenAI states what the genai client said about a call, under the names it said them.
//
// Nothing is added, subtracted or renamed here, and that is the point. This library runs inside somebody else's agent, so an interpretation it bakes into a number can only be corrected by every adopter upgrading and redeploying; one the server makes can be corrected by one deploy of ours, and can be reapplied to records already written because what the provider actually said is still on them.
//
// Quantities reported as zero are left out. A provider reporting nothing and a provider reporting none are the same thing to a bill.
func ReportedFromGenAI(usage *genai.GenerateContentResponseUsageMetadata) []*pb.ReportedQuantity {
	if usage == nil {
		return nil
	}

	reported := make([]*pb.ReportedQuantity, 0, 8)
	add := func(unit string, quantity int32) {
		if quantity != 0 {
			reported = append(reported, &pb.ReportedQuantity{Unit: unit, Quantity: int64(quantity)})
		}
	}

	add("promptTokenCount", usage.PromptTokenCount)
	add("candidatesTokenCount", usage.CandidatesTokenCount)
	add("cachedContentTokenCount", usage.CachedContentTokenCount)
	add("thoughtsTokenCount", usage.ThoughtsTokenCount)
	add("toolUsePromptTokenCount", usage.ToolUsePromptTokenCount)
	add("totalTokenCount", usage.TotalTokenCount)

	// The modality breakdowns, which are the only place it is visible that some of these tokens were audio or image rather than text. They are priced at their own rates, several times the text rate on both providers that report them, so a call that carries them and says so is the difference between a voice agent's bill being right and being a guess.
	addModalities(add, "promptTokensDetails", usage.PromptTokensDetails)
	addModalities(add, "candidatesTokensDetails", usage.CandidatesTokensDetails)
	addModalities(add, "cacheTokensDetails", usage.CacheTokensDetails)
	addModalities(add, "toolUsePromptTokensDetails", usage.ToolUsePromptTokensDetails)

	if len(reported) == 0 {
		return nil
	}
	return reported
}

// addModalities names each modality's share of a count, e.g. `candidatesTokensDetails.AUDIO`.
//
// A breakdown of a count already reported rather than a count of its own, which is why the parent name is kept as a prefix: whatever reads these has to be able to tell a part from a whole, and the name is the only thing that says which it is.
func addModalities(add func(string, int32), parent string, counts []*genai.ModalityTokenCount) {
	for _, count := range counts {
		if count == nil {
			continue
		}
		modality := strings.TrimSpace(string(count.Modality))
		if modality == "" {
			continue
		}
		add(parent+"."+modality, count.TokenCount)
	}
}

File: python/src/agentpulse/usage.py
"""What a provider said a call used, under the provider's own names.

Nothing is added, subtracted or renamed here, and that is the point. Providers disagree about what their own headline numbers contain: Gemini's prompt count already includes the tokens served from cache, OpenAI's does too and counts reasoning inside its output figure besides. Splitting the counts into priced classes happens once, on the server, where a mistake is corrected by one deploy and reapplied to records already written. An interpretation baked in here could only be corrected by every adopter upgrading.

The names match what the Go recorder sends for the same call, because the server reads both with the same reader.
"""

from __future__ import annotations

from typing import Any, Mapping

from ._wire import ReportedQuantity

# The conventions a set of counts can arrive in. Which one decides how the numbers are read, and it is not the same question as who billed: Claude served through Vertex is billed by Google and reports Anthropic-shaped usage.
FORMAT_VERTEX = "VERTEX"
FORMAT_ANTHROPIC = "ANTHROPIC"
FORMAT_OPENAI_CHAT = "OPENAI_CHAT"
FORMAT_OPENAI_RESPONSES = "OPENAI_RESPONSES"
# Every provider's usage as LiteLLM restates it, in OpenAI's shape but with both the cache read and the cache write inside the input count.
FORMAT_LITELLM = "LITELLM"
# OpenAI's chat field names with Perplexity's own counts beside them, such as the citation tokens its search charges for.
FORMAT_PERPLEXITY = "PERPLEXITY"

# Who bills for a call, in the vocabulary the rate card prices against. A plain string on the wire, so calling a provider nobody has called before needs no release of this library.
PROVIDER_VERTEX_AI = "VERTEX_AI"
PROVIDER_ANTHROPIC = "ANTHROPIC"
PROVIDER_OPENAI = "OPENAI"
PROVIDER_PERPLEXITY = "PERPLEXITY"

# Gemini's counts, in the order the Go recorder states them, under the names the REST api gives them. The Python sdk spells them in snake case; the server reads the REST names.
_GENAI_COUNTS = (
    ("prompt_token_count", "promptTokenCount"),
    ("candidates_token_count", "candidatesTokenCount"),
    ("cached_content_token_count", "cachedContentTokenCount"),
    ("thoughts_token_count", "thoughtsTokenCount"),
    ("tool_use_prompt_token_count", "toolUsePromptTokenCount"),
    ("total_token_count", "totalTokenCount"),
)

# The per-modality breakdowns, which are the only place it is visible that some tokens were audio or image rather than text. Those are priced at their own rates, several times the text rate.
_GENAI_DETAILS = (
    ("prompt_tokens_details", "promptTokensDetails"),
    ("candidates_tokens_details", "candidatesTokensDetails"),
    ("cache_tokens_details", "cacheTokensDetails"),
    ("tool_use_prompt_tokens_details", "toolUsePromptTokensDetails"),
)


def reported_quantities(reported: Mapping[str, int] | None) -> list[ReportedQuantity]:
    """What a record carries for a set of counts: sorted by name, with zeros and blank names left out.

    Sorted because a record whose counts land in a different order on every call cannot be compared with another recording of the same call, and comparing them is how a convention gets checked.
    """
    if not reported:
        return []
    return [
        ReportedQuantity(unit=unit, quantity=int(quantity))
        for unit, quantity in sorted(reported.items())
        if unit.strip() and _whole(quantity) and quantity != 0
    ]


def reported_from_genai(usage: Any) -> dict[str, int]:
    """The counts on a Gemini or Vertex reply's usage metadata, from the `google-genai` sdk's object or from the REST reply's `usageMetadata`."""
    if usage is None:
        return {}
    reported: dict[str, int] = {}
    for snake, camel in _GENAI_COUNTS:
        value = _read(usage, snake, camel)
        if _whole(value) and value:
            reported[camel] = int(value)
    for snake, camel in _GENAI_DETAILS:
        for detail in _read(usage, snake, camel) or ():
            modality = _read(detail, "modality", "modality")
            modality = str(getattr(modality, "value", modality) or "").strip()
            count = _read(detail, "token_count", "tokenCount")
            if modality and _whole(count) and count:
                reported[f"{camel}.{modality}"] = int(count)
    return reported


def reported_from(usage: Any) -> dict[str, int]:
    """Every whole-number count on a provider's usage object, each under the provider's own name, with a nested count named by its path, e.g. `prompt_tokens_details.cached_tokens`.

    For OpenAI's and Anthropic's sdks, whose usage objects are the reply's JSON under the JSON's own names, and for a usage dictionary read off any reply. Text, flags and lists are left out; a count reported as zero is too, because a provider reporting nothing and one reporting none are the same thing to a bill.
    """
    counts: dict[str, int] = {}
    _walk(_as_mapping(usage), "", counts, depth=0)
    return counts


reported_from_openai = reported_from
reported_from_anthropic = reported_from


def _walk(mapping: Mapping[str, Any] | None, prefix: str, out: dict[str, int], depth: int) -> None:
    if mapping is None or depth > 4:
        return
    for name, value in mapping.items():
        if not isinstance(name, str):
            continue
        if _whole(value):
            if value:
                out[prefix + name] = int(value)
            continue
        nested = _as_mapping(value)
        if nested is not None:
            _walk(nested, f"{prefix}{name}.", out, depth + 1)


def _as_mapping(value: Any) -> Mapping[str, Any] | None:
    if value is None or isinstance(value, (str, bytes, int, float, list, tuple)):
        return None
    if isinstance(value, Mapping):
        return value
    dump = getattr(value, "model_dump", None)
    if callable(dump):
        try:
            dumped = dump()
        except Exception:
            return None
        return dumped if isinstance(dumped, Mapping) else None
    to_dict = getattr(value, "to_dict", None)
    if callable(to_dict):
        try:
            dumped = to_dict()
        except Exception:
            return None
        return dumped if isinstance(dumped, Mapping) else None
    return None


def _whole(value: Any) -> bool:
    if isinstance(value, bool):
        return False
    if isinstance(value, int):
        return True
    return isinstance(value, float) and value.is_integer()


def _read(source: Any, attribute: str, key: str) -> Any:
    if isinstance(source, Mapping):
        return source.get(key, source.get(attribute))
    return getattr(source, attribute, None)
